File: 프로그래머스/unrated/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.js
function heal(health, healing, current) {
    if (healing <= 0)
        return current;
    current = current + healing;
    if (current >= health)
        current = health;
    return current;
}

function solution(bandage, health, attacks) {
    // 붕대 정보 [시전시간, 1초당 회복량 x, 추가 회복량 y], 최대 체력, [공격 시간,피해량]
    let answer = 0; // 남은 체력
    // t * x 체력 회복 연속으로 붕대 감기 성공하면 t * x + y
    // 최대 체력이 존재해 현재 체력이 최대 체력보단 작음
    // 공격 당하면 취소 && 회복 x && 체력 다운 + 붕대감기 다시 사용해서 연속 성공시간 0
    // 0 이하면 죽음

    let time = 0;
    const maxHealth = health;
    
    while(attacks.length != 0)
    {
        const prev = time;
        const next = time + bandage[0];
        if (attacks[0][0] > prev && attacks[0][0] <= next) {
            const damage = attacks.shift();
            time = damage[0];
            const healing = (damage[0] - prev - 1) * bandage[1];
            health = heal(maxHealth, healing, health);
            health -= damage[1];
            if (health <= 0)
                return -1;
        } else {
            const healing = (bandage[0] * bandage[1]) + bandage[2];
            health = heal(maxHealth, healing, health);
            time += bandage[0];
        }
    }
    return health;
}

File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int solution(vector<int> bandage, int health, vector<vector<int>> attacks) {
    // t초 동안 붕대를 감으며 1 초마다 x 만큼 체력 회복
    // 연속으로 성공하면 + y
    // 몬스터에게 당하면 기술이 취소됨, 당하는 순간은 회복 x
    // 체력 0 이하면 정지 -1
    // 남은 체력 리턴
    
    // 시작시간 0 으로 초기화
    // 어택을 순회하며 시작시간이랑 비교
    // 비교시 체력 재조정
    
    int time = 0;
    int atk = 0;
    int cnt = 0;
    const int maxHealth = health;
    const int binding = bandage[0];
    const int healing = bandage[1];
    const int success = bandage[2];
    while (atk < attacks.size()) {
        time += 1;
        cnt += 1;
        if (attacks[atk][0] == time) {
            health -= attacks[atk][1];
            atk += 1;
            cnt = 0;
        }
        else { 
            health += healing;
            if (cnt == binding) {
                health += success;
                cnt = 0;
            }
        }
        
        if (health >= maxHealth) health = maxHealth;
        else if (health <= 0) {
            health = -1;
            break ;
        }
    }
    return health;
}